### ISSUE TITLE
Emotion 10 & performance optimizations best practices

### ISSUE THREAD

**Issue by tadeuszwojcik** [emotion-js/emotion#859]
- `emotion` version: `@emotion/core`: `0.13.0`    `@emotion/styled-base`: `^0.10.3`
- `react` version: 16.5

Hi, I'm building UI lib that uses emotion 10, with build-in support for theming sort of similar to https://github.com/jxnblk/styled-system in some aspects.

I was wondering what things I could do to perf-optimize it, but there is not a lot of docs/info about it, hence my question about sort of best practices how to squeeze most out of emotion :)

I have following structure for theme object:

```javascript
export const theme = {
  colors: {
    primary: '#eef0f2',
    secondary: '#6ec6ff'
    ...
  },

  breakpoints: {
    xs: 320,
    sm: 600
   ....
  },

  fontSizes: {
    xs: 12,
    sm: 14
   ...
  },

  components: {
    button: {
      bg: 'secondary',
      fontSize: 'sm',
      css: ({ theme }) => ({
        ':hover': {
          backgroundColor: theme.colors.secondary
        }
      }),

      variants: {
        secondary: {
          bg: 'primary',

          css: ({ theme }) => ({
            ':hover': {
              backgroundColor: theme.colors.primary
            }
          })
        }
      }
    }
  }
};
```
I create components via custom factory that plugs into that theme structure, so can write:
```jsx
const Button = myFactory({
    tag: 'button',
    defaultTheme: {
     button: {
        bg: 'primary'
        ....
     }
});

<Button>Gets styles from theme.components.button</Button>
<Button variant="secondary">Gets styles from theme.components.button & its variant secondary</Button>
<Button bg="primary">Can override bg inline and primary get's replaced with it's value in theme.colors</Button>
<Button  css={{textTransform: 'uppercase'}}>Can add custom css via inline css as well</Button>
```
As you can see I need to potentially merge/override many css objects from many different places.

`themed props` also need to be transformed into normal css first, so for example `bg="primary"` gets transformed to `background: #eef0f2` .

Additionally any prop in css object or any themed prop (that get's it's value from theme, like `bg`), can have following `responsive` structure so I can write for example:
```jsx
<Button bg={{xs: 'primary', md: 'secondary'}}>Gets styles from theme.components.button</Button>
```
and it will be replaced with `min-width` media queries in resulting css object, so for this example it would be:
```css
.button-class {
   @media screen and (min-width: 320px) {
     background: #eef0f2
    }
   @media screen and (min-width: 600px) {
     background: #6ec6ff
    }
}
```

- Currently I do all of that merging in single function passed to `styled(tag)(myTransformCssFunc)` and merge css objects on my own (also execute any `css` prop function if it's a function not an object), but I was wondering if it perhaps would be more of best practice to pass many functions to styled so I'd had: `styled(tag)(transformDefaultThemeProps, TransformInlineProps, TransformInlinCssProp and so on)` , would that potentially be faster given that in many cased passed css objects would be or could be the same (same reference to object),  or it's best/most performant to do merging once so `@emotion` simply get's one object that only needs to be converted , not array of 8-10 objects that get merged inside `emotions` logic? I've seen before emotion 10, there was some kind of weak map memoization for Css objects getting converted to strings, but couldn't find it in v10, it's not applicable anymore or handled differently?

- Are there any tricks, patterns to get best possible performance when passing Css objects to emotion not strings? Ideally I'd like to cache as much as possible, but perhaps best practice would be to use `PureComponents` somewhere higher in the components hierarchy instead? I could perhaps also cache my custom transformCss function results so passed to emotion object(s) would have the same reference, but currently it would still `stringify` it every time?

- I use css object/function prop on my components, but it's not `@emotion` `css` prop with it's custom JSX factory. I assume that babel plugin won't hoist such props if those are `pure` as it only works for emotion css prop ?

I'm enjoying working with emotion 10 a lot, just want to make sure I use it in most optimized way, looked at the code a little bit, but I'm not sure If I get everything correctly.
I'd appreciate an tips!
Thanks!

**Comment by Andarist** [comment]
In general, as you know - reading by your comments, it all boils down to efficient memoization. If you can preserve object references then most of the things should just get retrieved from the cache.

I'm not familiar with current emotion@10 codebase, especially that it's still changing often~, so can't give you a concrete advise - the goal for emotion is definitely to cache as much as possible, so I would just work on that assumption. 

If you want to really squeeze out the most you'd have to debug your cases to see how exactly emotion handles them. 

> I use css object/function prop on my components, but it's not @emotion css prop with it's custom JSX factory. I assume that babel plugin won't hoist such props if those are pure as it only works for emotion css prop ?

Not sure about this (try it on a sample file 😉 ), i.e. `styled` calls are rewritten no matter what (unless u configure the plugin differently), even if they don't come from react-emotion package, so the same(-ish) might be true for `css` prop.

If you can make some work at compile-time this also might give you some perf boosts.

But generally speaking - don't try to overoptimize things unless you need to 😉

**Comment by mitchellhamilton** [comment]
> I use css object/function prop on my components, but it's not @emotion css prop with it's custom JSX factory. I assume that babel plugin won't hoist such props if those are pure as it only works for emotion css prop?

It will transform the css prop if there's an import like `import { jsx } from '@emotion/core'` or the `cssPropOptimization` option is enabled in the babel plugin. (in the babel plugin for emotion 10 that is on npm the option is named `jsx`)

> I've seen before emotion 10, there was some kind of weak map memoization for Css objects getting converted to strings, but couldn't find it in v10, it's not applicable anymore or handled differently?

That was primarily there because in v9 we would hoist objects in the babel plugin, instead in v10, they're hoisted and a css call is wrapped around it(it's different from the css function in v9, the one in v10 returns an object with the styles and the hash).

So basically, what you might want to do is if you have static objects, wrap them in a css call and use the result from the css call.

The main reason for not having the WeakMap cache in v10 is that it can't be completely safely used with styled because it breaks if an object that has a function inside it is cached.

WRT to the passing lots of objects vs one object to emotion, I'm not totally sure what would be better. The best thing to do is probably try both and see which is better but tbqh I doubt it will make a huge difference so unless it's causing a problem I would just leave it however you have it now.

**Comment by tadeuszwojcik** [comment]
Thanks guys, really appreciate it! @Andarist  I'ts not yet open source, but will be really soon. The idea is to have something similar'ish to rebassjs, but with bit different set of features & tradeoffs and written in TS.
